File: app/Packages/ONTFeatures/Sources/ReadingFeature/Presentation/SuiviDeLecture.swift
import SwiftUI

/// Ce que la liseuse sait de l'endroit où le lecteur en est.
///
/// ## Pourquoi une classe, et pourquoi elle n'est pas observable
///
/// Elle est écrite à chaque verset qui entre ou sort de l'écran, c'est-à-dire
/// des dizaines de fois par seconde pendant un défilement. Observable, elle
/// redéclencherait le corps de la liseuse à chaque fois — pour un
/// renseignement dont l'affichage n'a aucun besoin. C'est une ardoise, pas un
/// état de vue.
///
/// ## Le défaut qu'elle corrige
///
/// La position était écrite **directement** au modèle, sur changement de
/// visibilité. Or la visibilité change aussi quand la vue arrive et quand elle
/// part : pendant qu'une unité se retire, sa liste se défait et les lignes du
/// haut se signalent visibles une dernière fois. La position enregistrée
/// redevenait « verset 1 », et le défaut ne se voyait qu'à la visite suivante —
/// la première marchait, la seconde ramenait en haut du chapitre.
///
/// Relevé sur l'appareil, dans `lecteur.json` : `"verse": 1` après une lecture
/// qui s'était arrêtée bien plus bas.
///
/// Deux gardes s'y sont succédé et n'ont pas tenu : éteindre le suivi pendant
/// la restauration, puis l'éteindre aussi à la disparition. Toutes deux
/// supposaient de connaître l'**ordre** dans lequel SwiftUI émet ses
/// événements. On ne le connaît pas, et il n'est pas garanti.
///
/// ## Le signal qu'on écoute désormais
///
/// Non plus « une ligne est apparue », mais « le lecteur a fait défiler, et il
/// s'est arrêté ». Une apparition ni une disparition ne produisent ce
/// signal-là : le défaut disparaît sans qu'on ait à deviner un ordre.
@MainActor
final class SuiviDeLecture {
    /// Les versets présents à l'écran. Le plus petit est celui du haut.
    private var visibles: Set<Int> = []
    /// Vrai dès que le doigt a touché la page.
    private var aDefile = false

    func entre(_ verset: Int) { visibles.insert(verset) }
    func sort(_ verset: Int) { visibles.remove(verset) }
    func defile() { aDefile = true }

    /// Le verset à retenir — `nil` tant que le lecteur n'a rien fait défiler.
    ///
    /// Ouvrir une unité, la lire sans bouger et la quitter ne déplace donc pas
    /// la position : elle reste là où la restauration l'avait posée, ce qui est
    /// exactement ce qu'on veut.
    func aRetenir() -> Int? {
        guard aDefile else { return nil }
        return visibles.min()
    }

    /// À l'ouverture d'une autre unité.
    func recommence() {
        visibles.removeAll()
        aDefile = false
    }
}

private struct SuiviDeLectureKey: @preconcurrency EnvironmentKey {
    @MainActor static let defaultValue = SuiviDeLecture()
}

extension EnvironmentValues {
    /// L'ardoise du suivi de lecture.
    ///
    /// Passée par l'environnement pour n'avoir pas à la faire traverser cinq
    /// signatures de vues qui n'en font rien d'autre que la transmettre. Elle
    /// n'est pas observable, donc la lire ne crée aucune dépendance.
    var ontSuivi: SuiviDeLecture {
        get { self[SuiviDeLectureKey.self] }
        set { self[SuiviDeLectureKey.self] = newValue }
    }
}

File: android/ontfeatures/src/main/kotlin/com/labibleont/ont/features/reading/SuiviDeLecture.kt
package com.labibleont.ont.features.reading

import androidx.compose.runtime.Stable

/**
 * Ce que la liseuse sait de l'endroit où le lecteur en est.
 *
 * ## Ce qu'il répare
 *
 * La position de reprise n'était écrite qu'au **toucher** d'un verset — donc en
 * le sélectionnant. Qui lit en faisant défiler, sans jamais rien désigner, ne
 * déplaçait jamais sa reprise : « Reprendre » pointait le dernier verset touché,
 * parfois d'une tout autre séance. C'était l'usage principal, pas un cas limite.
 *
 * ## Pourquoi au verset et non au bloc
 *
 * Un premier correctif retenait le premier verset du **bloc** en tête d'écran,
 * parce que le bloc est l'élément de liste. En prose continue, un bloc est une
 * section entière dans un seul `Text` : la reprise ramenait alors au début de la
 * section au lieu de l'endroit qu'on lisait. iOS a connu exactement ce défaut et
 * l'a réparé — tant que les blocs valaient un verset, il ne se voyait pas, et
 * c'est la fusion en prose qui l'a révélé.
 *
 * ## Là où Android peut faire mieux qu'iOS, et le fait
 *
 * iOS estime la part de hauteur de chaque verset **au prorata des signes
 * affichés**, faute de pouvoir demander sa mise en page au moteur de texte : une
 * approximation qui vaut à un verset près, et qui se trompe entre un verset
 * serré et un verset aéré.
 *
 * Compose rend cette mise en page dans `TextLayoutResult`. On lit donc les
 * bornes **réelles** de chaque verset au lieu de les estimer. Le résultat est le
 * même à l'usage, il est simplement exact.
 *
 * ## La garde d'ouverture
 *
 * Reprise d'iOS telle quelle : tant que le lecteur n'a rien fait défiler,
 * [aRetenir] rend `null`. Ouvrir une unité, la lire sans bouger et la quitter ne
 * déplace donc pas la position — elle reste là où la restauration l'avait posée,
 * ce qui est exactement ce qu'on veut.
 */
@Stable
public class SuiviDeLecture {

    /**
     * Les bornes verticales de chaque verset, en coordonnées de la racine.
     *
     * ## Une carte ordinaire, et il faut que ça le reste
     *
     * Ni `mutableStateMapOf`, ni aucun état observable : rien ici ne doit
     * déclencher de recomposition. [situer] est appelée à chaque mise en page
     * de chaque bloc visible, c'est-à-dire très souvent pendant un défilement.
     *
     * Rendre cette carte observable coûterait une recomposition par entrée et
     * par sortie de verset — le prix se paierait par image, sur le geste le
     * plus courant de l'app. Le défilement vient tout juste d'être ramené à
     * 10 ms par image ; ce serait le premier endroit où il repartirait.
     *
     * iOS a exactement la même fragilité, pour la même raison : son
     * `SuiviDeLecture` est une classe ordinaire et non `@Observable`, et son
     * calcul de parts — 0,27 ms pour trente versets — ne tourne aujourd'hui
     * qu'à l'évaluation du corps. Le rendre observable les facturerait par
     * image des deux côtés.
     */
    private val bornes = sortedMapOf<Int, ClosedFloatingPointRange<Float>>()

    private var aDefile = false

    /** Le lecteur a touché au défilement — la position peut désormais bouger. */
    public fun defile() {
        aDefile = true
    }

    /** Où se trouve ce verset à l'écran. Appelé à chaque mise en page. */
    public fun situer(verset: Int, haut: Float, bas: Float) {
        if (bas > haut) bornes[verset] = haut..bas
    }

    /**
     * Ce verset n'est plus composé — on ne sait plus où il est.
     *
     * Sans cet oubli, ses bornes se figent à leur dernière valeur connue, qui
     * est **dans** la fenêtre puisque c'est là qu'il était quand on l'a mesuré
     * pour la dernière fois. Il resterait donc « visible » pour toujours, et
     * comme on retient le plus petit numéro, le verset 1 gagnerait à chaque
     * fois — quel que soit l'endroit où le lecteur est descendu.
     *
     * iOS n'a pas ce problème : sa sonde signale l'entrée **et la sortie** du
     * champ. Ici, la sortie du champ n'est pas un événement — c'est l'absence
     * d'événements. Il faut donc la prendre là où elle se manifeste : quand le
     * composable est mis au rebut.
     */
    public fun oublier(versets: Iterable<Int>) {
        versets.forEach(bornes::remove)
    }

    /** À l'ouverture d'une autre unité. */
    public fun recommence() {
        bornes.clear()
        aDefile = false
    }

    /**
     * Le verset à retenir — `nil` tant que le lecteur n'a rien fait défiler.
     *
     * Le plus petit numéro dont au moins la moitié est dans la fenêtre, comme le
     * seuil d'iOS. Un verset qui chevauche le bord haut sans y être à moitié
     * n'est pas celui qu'on lit : c'est celui qu'on vient de quitter.
     */
    public fun aRetenir(fenetreHaut: Float, fenetreBas: Float): Int? {
        if (!aDefile) return null
        return bornes.entries.firstOrNull { (_, borne) ->
            val hauteur = borne.endInclusive - borne.start
            val visible = minOf(borne.endInclusive, fenetreBas) - maxOf(borne.start, fenetreHaut)
            hauteur > 0f && visible / hauteur >= SEUIL
        }?.key
    }

    private companion object {
        /** Le seuil d'iOS — `onScrollVisibilityChange(threshold: 0.5)`. */
        const val SEUIL = 0.5f
    }
}
